### ISSUE TITLE
Possible sparse gradients bug in 1.4

### ISSUE THREAD

**Issue by nikonikolov** [tensorflow/tensorflow#14921]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Slackware 14.2
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: v1.4.0-rc1-11-g130a514 1.4.0
- **Python version**: 3.6

### Problem
I get the following warning when I run my code:
```
/usr/lib64/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py:96: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "
```
When I was using the exact same code with TF 1.3, I did not get the warning. I was wondering if this is a possible bug in 1.4 or simply the warning got added in the update?

I located the problem to this part of my code:

```
act_t_ph = tf.placeholder(tf.int32, [None])
# ...
with tf.variable_scope("action_value"):
    x = tf.layers.dense(x, 512, activation=tf.nn.relu)
    x = tf.layers.dense(x, 6, activation=None)
z_net_locs = x
action_mask = tf.one_hot(act_t_ph, 6, on_value=True, off_value=False, dtype=tf.bool)
z_locations = tf.boolean_mask(z_net_locs, action_mask)
# ...
```
Later, gradients are computed with respect to the variables in the dense layers and these gradients are backpropagated through  `z_locations`. 

I also tried changing my code to:

```
self.act_t_ph = tf.placeholder(tf.int32, [None])
# ...
with tf.variable_scope("action_value"):
    x = tf.layers.dense(x, 512, activation=tf.nn.relu)
    x = tf.layers.dense(x, 6, activation=None)
z_net_locs = x
bsize = tf.shape(self.act_t_ph)[0]
b_ind = tf.range(0, bsize, 1, tf.int32)
ind   = tf.concat([act_t_ph, b_ind], axis=-1)
z_locations = tf.gather_nd(z_net_locs, ind)
# ...
```
and I no longer get the warning in TF 1.4. As far as I can tell, one of the operations in the original code cannot handle sparse gradients.

**Comment by tatatodd** [comment]
Looks like that warning has existed for over a year:
https://github.com/tensorflow/tensorflow/blame/master/tensorflow/python/ops/gradients_impl.py#L96

But note that it's only a warning.  Is there actually anything wrong, other than the warning being generated?

**Comment by nikonikolov** [comment]
The code runs fine, but it is hard to tell if there is difference in speed, compared to TF 1.3.

I know that the warning has been there for a long time, but could it be that the implementation of some of the operators that I use changed between the two versions and causes the implicit conversion and hence the warning? In any case, it is very strange that I was not getting the warning in 1.3 if nothing changed.

**Comment by tatatodd** [comment]
@nikonikolov it's defnitely possible that something has changed, but again, note that it's just a warning.

I'm closing this out.  If you feel there's a concrete bug here, feel free to ping this issue (or open a new one), with a minimum reproducible example that demonstrates the issue, and I'll re-open.
